Expose bridge IPv6 setting to `docker network inspect`

aidanhs opened the pull request:
Using `docker network inspect bridge`.

Before:

```
        "options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500"
        }
```

After:

```
        "options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500",
            "com.docker.network.enable_ipv6": "false"
        }
```

aidanhs commented:
I think this is @mavenugo related?

aboch commented on daemon/daemon_unix.go:
```diff
@@ -419,6 +419,7 @@ func initBridgeDriver(controller libnetwork.NetworkController, config *Config) e
 		netlabel.DriverMTU:        strconv.Itoa(config.Mtu),
 		bridge.EnableIPMasquerade: strconv.FormatBool(config.Bridge.EnableIPMasq),
 		bridge.EnableICC:          strconv.FormatBool(config.Bridge.InterContainerCommunication),
+		netlabel.EnableIPv6:       strconv.FormatBool(config.Bridge.EnableIPv6),
```
`netOption` is really a collection of per network specific options. We consider  `EnableIPv6` a driver specific option instead, which is why it is being passed down separately from the network options.

aidanhs replied:
I don't understand the distinction. Unless I've misread the libnetwork code, it all ends up as properties on the network struct...so I'm not sure what it means to be 'driver specific'.

Regardless, this is just making the default bridge construction be more similar to manually created networks (for consistency).

aidanhs replied:
To elaborate more:

```
root@lp01679:/usr/bin# docker network create -o com.docker.network.enable_ipv6=true newbridge
13fc7ada2827d349d21a2c32fdd5a6fec32b6c0b8f591b7ec80d1dc4156146e1
root@lp01679:/usr/bin# docker network inspect newbridge
        "options": {
            "com.docker.network.enable_ipv6": "true"
        }
    }
]
root@lp01679:/usr/bin# ip addr
[...]
394: br-13fc7ada2827: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN 
    link/ether 02:42:8b:20:12:de brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-13fc7ada2827
       valid_lft forever preferred_lft forever
    inet6 fe80::1/64 scope link tentative 
       valid_lft forever preferred_lft forever
root@lp01679:/usr/bin# docker network rm newbridge
root@lp01679:/usr/bin# docker network create -o com.docker.network.enable_ipv6=false newbridge
410ec2597a8ed01125fc113ab1c51c37e55d481df2d82a50bd731b95702cfa0f
root@lp01679:/usr/bin# docker network inspect newbridge
[...]
        "options": {
            "com.docker.network.enable_ipv6": "false"
        }
    }
]
root@lp01679:/usr/bin# ip addr
[...]
395: br-410ec2597a8e: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN 
    link/ether 02:42:71:ff:88:ed brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-410ec2597a8e
       valid_lft forever preferred_lft forever
```

i.e. you can detect ipv6 on a network by looking at labels. It doesn't make sense to me that the default bridge would be a special case.

In fact, this is the only place in Docker where `NetworkOptionGeneric` is used! Although some of the labels might partition into driver-specific options, network specific options and generic options, that's a decision for the libnetwork internals to make.

Of course, I've only been playing around libnetwork for a day, so I may be misunderstanding.

mavenugo replied:
@aidanhs apologies on the delayed response. As you know libnetwork doesn't work on driver specific options/labels. Driver options are transparently sent to the drivers/plugins. Whereas libnetwork handles some of the generic labels such as `enable_ipv6` to provide driver independent functionality. In this case, `enable_ipv6` is used to update DNS records. That is the history behind keeping the enable_ipv6 flag as a driver independent flag (but still send it to the driver after libnetwork processing).
Having said that, I think there is an issue for user-defined networks wherein libnetwork doesn't process the this flag. I will confirm.

@mrjana can add more details to this.

LK4D4 commented:
ping @mavenugo @aboch

crosbymichael commented:
ping @mavenugo @aboch @mrjana yes/no/idk?

aidanhs commented:
Maybe some commentary will help clarify this PR:

1. the original intent was the desire to see the ipv6 enabled status as part of the default bridge settings, as you can for custom networks (as demonstrated in https://github.com/docker/docker/pull/17513#discussion_r43495336). Looking at the libnetwork [design document](https://github.com/docker/libnetwork/blob/master/docs/design.md#cnm-objects), this seems to mean turning the "option" into a "label".
2. as part of looking around the code, I observed that the default bridge is constructed in a different way to manually created networks (`CreateNetwork` in `daemon/network.go`). This motived the change in this PR to use `NetworkOptionDriverOpts`, which also fixes point 1.

If the reviewer believes that consistency is a good thing (i.e. agrees with these two points) then I think this PR is a step forward.

From a libnetwork design point of view, I'm not convinced that consumers of the library should be worrying about what kind of option something is, or how it's implemented behind the scenes - you just know that you want a network with setting S set to a value V. As things currently stand, libnetwork doesn't offer any guidance for choosing what type of option is appropriate (nor is anything enforced, nor are different types of options treated differently - both `NetworkOptionDriverOpts` and `NetworkOptionGeneric` are aware of the ipv6 option). Following this PR, I would therefore like to discuss (on the libnetwork repo) what value is provided by having these different types of options, because the topic is not touched on in the design document.

thaJeztah commented:
ping @mavenugo @aboch @mrjana could you have a look, and provide some input? (or a (not)LGTM)

@aidanhs looks like this needs a rebase (just something I noted, but no mayor issue, because it's still in design review)

tiborvass commented:
Ping @mavenugo @mrjana

mavenugo commented:
@tiborvass @aidanhs as explained in https://github.com/docker/docker/pull/17513#discussion_r46907343, this is a libnetwork options processing design especially providing a way for driver specific vs driver independent options. We had a few issues with this in the past (due to the labels vs options support).
@mrjana will have more thing to say on this topic.

aidanhs committed:
Expose bridge IPv6 setting to `docker network inspect`

Signed-off-by: Aidan Hobson Sayers <aidanhs@cantab.net>

aidanhs commented:
> this is a libnetwork options processing design especially providing a way for driver specific vs driver independent options

Just so I'm completely clear, you're saying that the split between driver specific and driver independent options (i.e. forcing them to be two different parameters to libnetwork) was a decision made to resolve some issues related to labels vs options?

The problem is that library consumers (like docker) now need to keep their own record of which options are driver specific/independent. This issue (`com.docker.network.enable_ipv6` only appearing in `docker network inspect` for user-defined networks) is caused by exactly that - someone didn't partition the options beforehand and so all options defined by the user are passed as 'driver-specific' options.

> Having said that, I think there is an issue for user-defined networks wherein libnetwork doesn't process the this flag. I will confirm.

As it happens the flag *is* actually processed for user networks, despite their creation *not* conforming to the intended design (where options should be split before being passed) - https://github.com/docker/docker/blob/3d82564c7a6ae139879b869a987bf9274ed0275d/daemon/network.go#L103. I assume this is a (happy) accident - when I read the code a while ago, it seemed to treat driver-specific and driver-independent options in the same way, it just had a list of settings it would look out for.

You can see some evidence of this working by comparing the output of `ip addr`:
```
root@lp01679:~# docker network create -o com.docker.network.enable_ipv6=false x
149adf318ef59feb4eb1d8d6bd2ca1fb4e1eb177cdff9963457a9b8cd9a9aeff
root@lp01679:~# docker run --net x --rm ubuntu:14.04.3 ip addr show eth0
67: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link 
       valid_lft forever preferred_lft forever
root@lp01679:~# docker network rm x
root@lp01679:~# docker network create -o com.docker.network.enable_ipv6=true x
d1ea63fc81e7c26f8ad63114b3839bb746dda84c4ad32396b72e48730ee2e0d8
root@lp01679:~# docker run -it --net x --rm ubuntu:14.04.3 ip addr show eth0
62: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:2/64 scope link 
       valid_lft forever preferred_lft forever
    inet6 fe80::242:ac12:2/64 scope link 
       valid_lft forever preferred_lft forever
```
Note the extra inet6 address for the latter.

Just as a reminder of the issue I actually want fixed: `com.docker.network.enable_ipv6` should be consistently displayed (or not displayed) for both the default bridge and user-created networks. At this point I really don't mind how it's resolved, I just think consistency is good. Perhaps your preferred solution (assuming you like the driver independant/dependant split) is to add some code in libnetwork to recombine everything into a consistent view of labels to display to the user.

I appreciate your responses and look forward to the response from @mrjana, but if you're sure that you're not going to merge the PR in its current state then I encourage you to just close it - I will raise an appropriate issue against docker and you can decide how to resolve it in line with your preferred design.

thaJeztah commented:
ping @mrjana can you have a look at this one please? thanks!

mrjana commented:
@aidanhs @mavenugo Let me give some context on what we actually want to do here regardless of what state the current implementation is in. Before 1.9 we only had default bridge and hence all the optional flags that were meant for the bridge driver were either just provided as a daemon option if it applies to all containers using bridge or in `docker run` if it is container specific. 

One such daemon option is `--ipv6` which is introduced to enable ipv6 connectivity in default bridge network. But with the introduction of multiple networks this should really be a per-network option except that  for backward compatibility we cannot really remove that daemon option. 

So the decision we made is that all the existing daemon wide networking options which existed pre-1.9 will only be applicable to default bridge network(with the exception of some options which are truly global and will apply to all bridge networks)

For all the other non-default bridge networks these options have to specified during the network create time. So in theory we should have `--ipv6` flag available as part of `network create` command and ipv6 connectivity should only be enabled in that network only if that flag is present. Not as it is currently done today where the daemon level `--ipv6` flag is applied on all networks.

Once we have the `--ipv6` option in the network command as a first class option, then there is no need to expose the label (as we do it today) for any network since it should show up as a first class configuration info for that network. We may still pass this information down to the driver in some arbitrary way but that doesn't need to be exposed in the UI.

So to me the design roughly should like this:

* Add a `--ipv6` option at the network level. It is for any network and if the backing driver cannot support ipv6 it has the option to reject the configuration and hence result in network creation failure.
* Remove accepting ipv6 configuration via labels
* Remove listing this as part of labels list in `network inspect` for any network
* Add a field in network inspect to show the first class ipv6 configuration for every network including default bridge network.

Thoughts?

mavenugo commented:
@mrjana thanks for the clarification and your suggestion make sense to me.

aidanhs commented:
> So the decision we made is that all the existing daemon wide networking options which existed pre-1.9 will only be applicable to default bridge network(with the exception of some options which are truly global and will apply to all bridge networks)

Makes sense.

> So to me the design roughly should like this:
> - Add a --ipv6 option at the network level. It is for any network and if the backing driver cannot support ipv6 it has the option to reject the configuration and hence result in network creation failure.
> - Remove accepting ipv6 configuration via labels
> - Remove listing this as part of labels list in network inspect for any network
> - Add a field in network inspect to show the first class ipv6 configuration for every network including default bridge network.

This is a way forward and it's better from the status quo, so +1 from that point of view.

However, my personal vague feeling is that labels are great! Why create another way of passing options down to drivers when you have an extensible, decent method already available to you? .ini files are just a string -> string set of keys and values, and they work pretty well for general configuration - I'm not sure what making ipv6 a first-class option buys you. If you like `--ipv6`, couldn't you use it as sugar for defining the label?

Or is it about the conceptual difference between labels and options (which I've not yet quite grasped)?

mrjana commented:
@aidanhs I think the term `labels` has been used to mean many things. But in docker it is a means to assign a tag/alias to a particular object whether it be an image or container or may be in the future to networks etc. What we really have in libnetwork is driver options(I guess the usage of the term netlabel there is the source of confusion) and as the name says they are driver specific options which only make sense for a particular driver and as such these should only be used to specify driver specific parameters.

If there is some option which is driver independent, then that should be really a first class option in docker api and cli. `--ipv6` belongs in that category and that is the reasoning behind what design I suggested.

If you agree with the above rough sketch, can you please rework the PR to address the design or close this one and create another one which addresses the problem the correct way?

icecrime commented:
Ping @aidanhs! Thanks :-)

aidanhs commented:
I see, so you're coming at it from the point of view that the number of driver-independent options is finite (and small), so they can all be supported as first-class values without asking too much of libnetwork consumers to keep track of them.

I'm not sure I 100% agree with the proposed route (and feel that you may end up redesigning how options work), but I understand the justification and can see it's an improvement so I'll make the required changes to this PR so forward progress can be made. People actually using libnetwork can then give feedback in the future for incremental improvement.

While I'm making the changes, I'm curious - what is the significance of the unique `com.docker.network.driver.mtu` name? Driver-specific options appear to replace `driver` with the name of a driver (like `bridge`) and driver-independent options simply omit `driver` entirely. Is mtu a special case?

mrjana commented:
@aidanhs Thanks for moving this forward. 

`com.docker.network.driver.mtu` also falls in the same bucket as `ipv6`. It needs to become generic as it is applicable to all drivers. Once we do that we can remove this label as well.

thaJeztah commented:
ping @aidanhs what's the status on this one; are you still working on this?

aidanhs commented:
Rebased, created two additional required PRs in engine-api and libnetwork, referenced above (I won't belabour the point, but having to change three repositories feels like a herculean effort!).

Please take a look. The IPv6 setting is now exposed as `--ipv6` and is also returned by `docker network inspect` as `EnableIPv6`. `Internal` is also now returned by `docker network inspect` while I was there.

There are some tests I've changed to verify the changes work, but obviously jenkins won't work and I've not yet managed to run them. I'll see if I can do a full test run overnight.

aidanhs commented:
Ran the tests. All unit tests pass. All but three integration tests pass, and those are something to do with checking out of memory conditions.

icecrime commented:
This is not building.

```
18:20:04 # github.com/docker/docker/api/client
18:20:04 api/client/network.go:81: unknown types.NetworkCreate field 'EnableIPv6' in struct literal
18:20:52 # github.com/docker/docker/api/server/router/network
18:20:52 api/server/router/network/network_routes.go:95: create.EnableIPv6 undefined (type "github.com/docker/engine-api/types".NetworkCreate has no field or method EnableIPv6)
18:20:52 api/server/router/network/network_routes.go:164: r.EnableIPv6 undefined (type *"github.com/docker/engine-api/types".NetworkResource has no field or method EnableIPv6)
18:20:52 api/server/router/network/network_routes.go:164: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
```

aidanhs commented:
True! It needs libnetwork and engine-api versions bumping (because my PRs to those repos have been merged) before the build will work. Should I do that as part of this PR?

thaJeztah commented:
@aidanhs please do; as a separate commit

aidanhs commented:
I've had a poke around and I'm not really sure how this is meant to work. The required changes are in libnetwork and engine-api, as can be seen by the following merge commits:

https://github.com/docker/libnetwork/commit/e1bf6811656e0821e3e26e597dfc031e3168fd82
https://github.com/docker/engine-api/commit/374736c8b20447a94727e30d68ee948c047cd303

However, both of these merges are into master, rather than the release branch. What's the process for getting something into the release branch so it can be used for this PR? I was going to edit `hack/vendor.sh` to use the commitish of the merges linked above (just to demonstrate that the tests do pass), but this pulls in some unrelated incompatible changes sitting on master of engine-api, so I'd have to make more changes to handle this incompatibility, then undo them in time for merge because the incompatible change may not end up on the release branch of engine-api!

Do you have some kind of guidance for making this kind of change? Should my first step be to beg the maintainers of libnetwork and engine-api to cherry-pick my changes onto the release branches (with the argument that they should be backwards compatible)? If there's some kind of document I'm sure that'd be helpful for external contributors like myself.

jfrazelle commented:
```
---> Making bundle: binary (in bundles/1.10.0-dev/binary)
18:31:29 Building: bundles/1.10.0-dev/binary/docker-1.10.0-dev
18:31:51 # github.com/docker/docker/api/client
18:31:51 api/client/network.go:81: unknown types.NetworkCreate field 'EnableIPv6' in struct literal
18:32:37 # github.com/docker/docker/api/server/router/network
18:32:37 api/server/router/network/network_routes.go:95: create.EnableIPv6 undefined (type "github.com/docker/engine-api/types".NetworkCreate has no field or method EnableIPv6)
18:32:37 api/server/router/network/network_routes.go:164: r.EnableIPv6 undefined (type *"github.com/docker/engine-api/types".NetworkResource has no field or method EnableIPv6)
18:32:37 api/server/router/network/network_routes.go:164: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
18:32:38 Build step 'Execute shell' marked build as failure
18:32:40 [PostBuildScript] - Execution post build scripts.
18:32:42 [Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson7350230624918664321.sh
18:32:42 + set +x
18:32:42 
18:32:42 
```
ping @mavenugo @mrjana I want to make sure this is something you all want, this PR has been around awhile

mavenugo commented:
@jfrazelle @aboch & @mrjana are working with @aidanhs in https://github.com/docker/libnetwork/pull/916 regarding this issue.

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/875/console):

~~~console

---
nEnableIPv6
+ ec=1
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 1 at Wed Feb 10 23:08:20 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 23:08:21 CUT 2016 (1m 27s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson6538942339214143228.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21637/console):

~~~console

---
/docker/daemon
daemon\network.go:12: imported and not used: "github.com/docker/libnetwork/netlabel"
daemon\network.go:13: imported and not used: "github.com/docker/libnetwork/options"
daemon\network.go:109: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson9146814193264682364.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6099/console):

~~~console

---
nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:12: imported and not used: "github.com/docker/libnetwork/netlabel"
daemon/network.go:13: imported and not used: "github.com/docker/libnetwork/options"
daemon/network.go:109: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson5448743293554599052.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/876/console):

~~~console

---
0T23:09:46Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of cleanup
INFO: Ended at Wed Feb 10 23:09:46 CUT 2016 (1m 24s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson2293334049159197864.sh
+ set +e
+ set +x
INFO: Nuking /d/CI
time="2016-02-10T23:09:54Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of clean
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21638/console):

~~~console

---
ork
api\server\router\network\network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon\network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson3095186959687242858.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6100/console):

~~~console

---
dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/network
api/server/router/network/network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson2502347634133596946.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1775/console):

~~~console

---
er/docker/daemon
daemon/daemon_unix.go:697:14: error: reference to undefined identifier 'libnetwork.NetworkOptionEnableIPv6'
   libnetwork.NetworkOptionEnableIPv6(config.bridgeConfig.EnableIPv6),
              ^
daemon/network.go:107:43: error: reference to undefined identifier 'libnetwork.NetworkOptionEnableIPv6'
  nwOptions = append(nwOptions, libnetwork.NetworkOptionEnableIPv6(enableIPv6))
                                           ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson5345561314512867510.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23658/console):

~~~console

---
dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/network
api/server/router/network/network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson7568547725276743738.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14868/console):

~~~console

---
dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/network
api/server/router/network/network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson3715048911362503306.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1776/console):

~~~console

---
er/docker/daemon
daemon/daemon_unix.go:697:14: error: reference to undefined identifier 'libnetwork.NetworkOptionEnableIPv6'
   libnetwork.NetworkOptionEnableIPv6(config.bridgeConfig.EnableIPv6),
              ^
daemon/network.go:107:43: error: reference to undefined identifier 'libnetwork.NetworkOptionEnableIPv6'
  nwOptions = append(nwOptions, libnetwork.NetworkOptionEnableIPv6(enableIPv6))
                                           ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson4058376922010390134.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23657/console):

~~~console

---
dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/network
api/server/router/network/network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson7161165064488202807.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14867/console):

~~~console

---
dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/api/server/router/network
api/server/router/network/network_routes.go:163: nw.Info().IPv6Enabled undefined (type libnetwork.NetworkInfo has no field or method IPv6Enabled)
# github.com/docker/docker/daemon
daemon/daemon_unix.go:697: undefined: libnetwork.NetworkOptionEnableIPv6
daemon/network.go:107: undefined: libnetwork.NetworkOptionEnableIPv6
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson2437160289617127363.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/879/console):

~~~console

---
nEnableIPv6
+ ec=1
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 1 at Wed Feb 10 23:30:04 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 23:30:05 CUT 2016 (1m 28s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\jenkins\AppData\Local\Temp\hudson5602477575692281918.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/881/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.1 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.7%
 18.3%
 22.0%
 25.7%
 29.4%
 33.0%
 36.7%
 40.4%
 44.1%
 47.8%
 51.4%
 55.1%
 58.8%
 62.5%
 66.1%
 69.8%
 73.5%
 77.2%
 80.8%
 84.5%
 88.2%
 91.9%
 95.6%
 99.2%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.38 m   54.38 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1163904122 Bytes/sec.
   Speed :            66599.128 MegaBytes/min.
   Ended : Wednesday, February 10, 2016 11:33:31 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-4
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/156/console):

~~~console

---
7
Step 49 : COPY . /go/src/github.com/docker/docker
 ---> 8d6dffb97c0a
Removing intermediate container 452baeff7b87
Successfully built 8d6dffb97c0a

---> Making bundle: validate-vendor (in bundles/1.11.0-dev/validate-vendor)
The result of ./hack/vendor.sh differs

 M vendor/src/github.com/docker/libnetwork/default_gateway_linux.go
 M vendor/src/github.com/docker/libnetwork/network.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson7100196232589731247.sh
+ set +x


---

~~~

aidanhs commented:
(I think Gordon has forgotten to remove a message - https://github.com/docker/docker/pull/17513#issuecomment-182635884)

Ok, the PR referenced by @mavenugo has been merged, and I now have two commits on master of libnetwork (the one on engine-api has already made it into docker master)

 - https://github.com/docker/libnetwork/commit/e1bf6811656e0821e3e26e597dfc031e3168fd82
 - https://github.com/docker/libnetwork/commit/381061f184687732ec2f30877de5d81c1b31c308

In order to demonstrate that this PR will work when it's all pulled together, I've cherry-picked those two commits into this branch (hence the 'vendor' test failing) and you can see that all of the functionality tests have passed.

Answers to my questions in https://github.com/docker/docker/pull/17513#issuecomment-178902515 will inform my next steps - what is the correct way for me to request the libnetwork maintainers cherry-pick my changes into the release branch?

mavenugo commented:
@aidanhs  I dont think there is anything in libnetwork master that would require any special handling in docker/docker. Hence you can vendor it in yourselves.
But,  since it requires a new tagging mechanism, I will handle it now and vendor it in which will carry your changes.

mavenugo commented:
@aidanhs https://github.com/docker/docker/pull/20214 brings in your changes from libnetwork. once this is merged you can rebase this PR.
cc @icecrime

aidanhs commented:
Thanks @mavenugo!

calavera commented:
LGTM

aboch commented:
LGTM

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14917/console):

~~~console

---
ReadProcTimer	0.504s
PASS: docker_cli_run_unix_test.go:28: DockerSuite.TestRunRedirectStdout	0.989s
PASS: docker_cli_run_test.go:1389: DockerSuite.TestRunResolvconfUpdate	3.872s
PASS: docker_cli_run_test.go:2701: DockerSuite.TestRunRestartMaxRetries	2.729s
PASS: docker_cli_run_test.go:2147: DockerSuite.TestRunReuseBindVolumeThatIsSymlink	1.056s
PASS: docker_cli_run_test.go:1179: DockerSuite.TestRunRootWorkdir	0.524s

----------------------------------------------------------------------
FAIL: docker_cli_run_unix_test.go:842: DockerSuite.TestRunSeccompAllowAptKey

docker_cli_run_unix_test.go:848:
    c.Fatalf("expected apt-key with seccomp to succeed, got %s: %v", out, err)
... Error: expected apt-key with seccomp to succeed, got Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.fo0ZVM4KRp --no-auto-check-trustdb --trust-model always --primary-keyring /etc/apt/trusted.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-automatic.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-security-automat
---

~~~

calavera commented:
@aidanhs This is going to need docs.

aidanhs committed:
Add docs for --ipv6 option, also add --internal as appropriate

Signed-off-by: Aidan Hobson Sayers <aidanhs@cantab.net>

aidanhs commented:
I've just added docs as a followup commit for ease of review.

icecrime commented:
Thanks! Moving to docs review.

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14933/console):

~~~console

---
ent	0.979s
PASS: docker_cli_start_test.go:112: DockerSuite.TestStartMultipleContainers	1.379s
PASS: docker_cli_start_test.go:96: DockerSuite.TestStartPausedContainer	0.707s
PASS: docker_cli_start_test.go:71: DockerSuite.TestStartRecordError	1.066s
PASS: docker_api_containers_test.go:1433: DockerSuite.TestStartWithNilDNS	0.340s
PASS: docker_api_containers_test.go:890: DockerSuite.TestStartWithTooLowMemoryLimit	0.165s

----------------------------------------------------------------------
FAIL: docker_cli_stats_test.go:98: DockerSuite.TestStatsAllNewContainersAdded

docker_cli_stats_test.go:131:
    c.Fatal("failed to observe new container created added to stats")
... Error: failed to observe new container created added to stats


----------------------------------------------------------------------
PASS: docker_cli_stats_test.go:80: DockerSuite.TestStatsAllNoStream	2.893s
PASS: docker_cli_stats_test.go:56: DockerSuite.TestStatsAllRunningNoStream	3.256s
PASS: docker_cli_stats_test.go:43: DockerSuite.TestStatsContainerNotFound	3.081s
PASS: docke
---

~~~

vdemeester commented on docs/reference/commandline/network_create.md:
```diff
@@ -134,7 +135,13 @@ The following are those options and the equivalent docker daemon flags used for
 | `com.docker.network.bridge.enable_icc`           | `--icc`     | Enable or Disable Inter Container Connectivity        |
 | `com.docker.network.bridge.host_binding_ipv4`    | `--ip`      | Default IP when binding container ports               |
 | `com.docker.network.mtu`                         | `--mtu`     | Set the containers network MTU                        |
-| `com.docker.network.enable_ipv6`                 | `--ipv6`    | Enable IPv6 networking                                |
```
I'm not sure what removing this from there though ? It's still a valid option when creating a bridge network right ?

aidanhs replied:
The context is the paragraph above

>  When creating a custom network, the default network driver (i.e. `bridge`) has additional options that can be passed.  The following are those options and the equivalent docker daemon flags used for docker0 bridge:

`com.docker.network.enable_ipv6` is no longer a valid 'driver option' (also known as 'label') and it will be ignored, so I've removed this line. Instead, `--ipv6` is a valid option for *all* networks.

vdemeester commented:
@aidanhs There is a few update to do in the `docs/reference/api/docker_remote_api_v1.23.md` (https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.23.md#25-networks — mainly in the response) and also probably one in `docs/reference/api/docker_remote_api.md` for the changes (https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v123-api-changes)

/cc @thaJeztah

thaJeztah commented on docs/reference/commandline/network_create.md:
```diff
@@ -134,7 +135,13 @@ The following are those options and the equivalent docker daemon flags used for
+
+| Argument     | Equivalent | Description                              |
+|--------------|------------|------------------------------------------|
+| `--internal` | -          | Restricts external access to the network |
```
What does the "equivalent" column relate to here? Is that the *daemon* flag?

mavenugo replied:
Also, `docker network create` takes in few more flags that are applicable to any driver such as `subnet`, `ip-range`, `gateway` & others. They are all already documented in `docs/reference/commandline/network_create.md`. am not sure if we need to repeat it here.

aidanhs replied:
@thaJeztah yes, this is in the context of the table above which says "and the equivalent docker daemon flags used for docker0 bridge" above it.

@mavenugo true. Want me to remove them? If so, I'll add a sentence above the exiting table saying something like "There are also arguments to `docker network create` applicable to any network driver, which can be found on the reference page for that command."

mavenugo replied:
@aidanhs thanks. Actually, the table is infact useful for those who like to compare the existing daemon flags with the `network` flags. So, maybe you can include all the options and compare them ?
For example `--ip-range` is equivalent to `--fixed-cidr`, the `--subnet` is equivalent to the subnet in `bip` , etc.

thaJeztah replied:
We may want to add a small bit to explain "what" they're equivalent to (i.e. that the "equivalent" column shows the equivalent `docker daemon` option)?

thaJeztah commented:
Okay, we discussed this; to prevent more rebase hell, well merge this, and look at the final changes in a follow up

LGTM

calavera commented:
docs LGTM. Let's follow up with another PR to add more information to that table with other options.
